core/textinput/src/textinput/Editor.cpp

```diff
@@ -430,12 +464,20 @@ namespace textinput {
         return kPRSuccess;
       case kCmdReverseSearch:
         PushUndo();
-        fMode = kHistSearchMode;
-        fSearch.clear();
+        if (fMode != kHistFwdSearchMode) fSearch.clear();
+        fMode = kHistRevSearchMode;
         SetReverseHistSearchPrompt(R.fDisplay);
         fContext->GetKeyBinding()->EnableEscCmd(true);
         if (UpdateHistSearch(R)) return kPRSuccess;
         return kPRError;
+      case kCmdForwardSearch:
+         PopUndo();
+         if (fMode != kHistRevSearchMode) fSearch.clear();
+         fMode = kHistFwdSearchMode;
+         SetForwardHistSearchPrompt(R.fDisplay);
+         fContext->GetKeyBinding()->EnableEscCmd(true);
+         if (UpdateHistSearch(R)) return kPRSuccess;
+         return kPRError;
```

Comment:
After changing `SetForwardHistSearchPrompt()` to `SetHistSearchModePrompt()` the code for this case should be similar to `kCmdReverseSearch`.  Therefore, I'd suggest to merge both code paths as in:
```c++
case kCmdReverseSearch:
case kCmdForwardSearch:
  // ...
  fMode = (M == kCmdReverseSearch ? kHistRevSearchMode : kHistFwdSearchMode);
  // ...
  break;
```